### PR TITLE
[Merged by Bors] - feat(tactic_log.cpp): log tactic invocations in AST

### PR DESCRIPTION
**Pull request by digama0**
This augments the `--ast` option to generate files with tactic invocation data. The way it works is that `tactic.istep` calls the `with_ast` meta constant, which will log the tactic state before and after each tactic invocation, in addition to the AST node of the tactic itself. Tactic invocations are added to a file-global list, so they may end up reordered nondeterministically because of parallel proof processing, but the tactic calls for a single declaration should be in order as long as the `async` tactic is not used.

The type signature of `istep` is changed (to add an additional `ast` argument); this affects all tactic classes, including custom tactic classes, so this will break downstream libraries that use custom tactic classes (the fix is easy though, see `tests/lean/run/my_tac_class.lean`).

### CODE REVIEW

File: src/library/ast_exporter.cpp [hunk 1]
```diff
@@ -158,6 +159,33 @@ struct ast_exporter : abstract_ast_exporter {
             {"level", std::move(m_levels)},
```
Comment (dselsam) [inline src/library/ast_exporter.cpp:159]: These two lines need to be *after* the tactic log exporting.

File: src/library/tactic/tactic_log.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+tactic_state_id tactic_log::push_invocation(ast_id id, tactic_state_id start, tactic_state_id end, bool success) const {
+    lock_guard<mutex> l(m_mutex);
+    auto& invocs = get_invocs(l);
+    tactic_state_id n = invocs.size();
+    invocs.emplace_back(id, start, end, success);
+    return n;
+}
+
+void tactic_log::detach() const {
+    if (m_detached.exchange(true)) {
```
Comment (dselsam) [inline src/library/tactic/tactic_log.cpp:50]: Did you mean to negate this?

**Comment by bryangingechen** [comment]
bors try

**Comment by digama0** [comment]
bors try

**Comment by bryangingechen** [comment]
Is it worth adding some documentation for `--ast` in this repo (maybe even just a pointer to mathport)?

**Comment by digama0** [comment]
I will write some documentation as a `.md` file (in a separate PR).

**Comment by bryangingechen** [comment]
OK, let's merge this now then.
bors r+

**Pull request closed without merging**